mindmeld/query_cache.py

```diff
@@ -14,131 +14,197 @@
+    def get_key(domain, intent, query_text):
+        '''
+        Calculates a hash key for the domain, intent and text of an example.
+        This key is required for further interactions with the query cache.
 
         Args:
-            domain (str): The domain
-            intent (str): The intent
-            query_text (str): The query text
-            processed_query (ProcessedQuery): The ProcessedQuery \
-                object corresponding to the domain, intent and query_text
-        """
-        if (domain, intent, query_text) in self.cached_queries:
-            return
-
-        self.cached_queries[(domain, intent, query_text)] = processed_query
-        self.is_dirty = True
-
-    def get_value(self, domain, intent, query_text):
-        """
-        Gets the value associated with the triplet key (domain, intent, query_text).
-
+            domain(str): The domain of the example
+            intent(str): The intent of the example
+            query_text(str): The raw text of the example
+        Returns:
+            str: Hash id representing this query
+        '''
+
+        h = sha256(domain.encode())
+        h.update(b'###')
+        h.update(intent.encode())
+        h.update(b'###')
+        h.update(query_text.encode())
+        return h.hexdigest()
+
+    def key_to_row_id(self, key):
+        '''
         Args:
-            domain (str): The domain
-            intent (str): The intent
-            query_text (str): The query text
-        """
-        try:
-            return self.cached_queries[(domain, intent, query_text)]
-        except KeyError:
-            return
-
-    def dump(self):
-        """
-        This function dumps the query cache mapping to disk. This operation is expensive,
-        so use it sparingly!
-        """
-        if not self.is_dirty:
-            return
+            key(str): A key generated by the QueryCache.get_key() function
+        Returns:
+            Optional(Integer): Unique id of the query in the cache if it exists
+        '''
+
+        cursor = self.connection.cursor()
+        cursor.execute('''
+        SELECT rowid FROM queries where hash_id=(?);
+        ''', (key,))
+        row = cursor.fetchone()
+        return row[0] if row else None
+
+    def put(self, key, processed_query):
+        '''
+        Adds a ProcessedQuery to the cache
 
-        # make generated directory if necessary
-        if not os.path.isdir(self.gen_folder):
-            os.makedirs(self.gen_folder)
+        Args:
+            key(str): A key generated by QueryCache.get_key() for this example
+            processed_query(ProcessedQuery): The ProcessedQuery generated for this example
+        Returns:
+            integer: The unique id of the query in the cache.
+        '''
+
+        cursor = self.connection.cursor()
+        cursor.execute('''
+        INSERT OR IGNORE into queries values (?, ?, ?, ?, ?)
+        ''', (key,
+              json.dumps(processed_query.to_cache()),
```

Comment:
We could use ultrajson here (https://github.com/ultrajson/ultrajson), based on these benchmarks: https://medium.com/@shmulikamar/python-serialization-benchmarks-8e5bb700530b since it has a faster serialization/deserialization than JSON stdlib. But the performance is not a significant improvement, so maybe we can punt on it.

Reply:
My general opinion is that we need fairly significant reasons to bring in another dependency.  This doesn't seem like one since we'd probably not see a major difference in training time.